File: src/main/java/me/alegian/thavma/impl/common/block/CrucibleBlock.java
package me.alegian.thavma.impl.common.block;

import me.alegian.thavma.impl.common.aspect.AspectHelperKt;
import me.alegian.thavma.impl.common.aspect.AspectMap;
import me.alegian.thavma.impl.common.block.entity.CrucibleBE;
import me.alegian.thavma.impl.common.data.capability.AspectContainer;
import me.alegian.thavma.impl.common.data.capability.IAspectContainer;
import me.alegian.thavma.impl.common.recipe.CrucibleRecipeInput;
import me.alegian.thavma.impl.init.registries.T7Tags;
import me.alegian.thavma.impl.init.registries.deferred.T7BlockEntities;
import me.alegian.thavma.impl.init.registries.deferred.T7RecipeTypes;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.ItemInteractionResult;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.item.ItemEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.item.crafting.RecipeManager;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LevelAccessor;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.EntityBlock;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityTicker;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockBehaviour;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BooleanProperty;
import net.minecraft.world.level.pathfinder.PathComputationType;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.shapes.BooleanOp;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.phys.shapes.VoxelShape;
import net.neoforged.neoforge.fluids.FluidUtil;
import org.jetbrains.annotations.Nullable;

public class CrucibleBlock extends Block implements EntityBlock {
  public static final VoxelShape CRUCIBLE_INSIDE = Block.box(2.0, 4.0, 2.0, 14.0, 16.0, 14.0);
  public static final VoxelShape CRUCIBLE_SHAPE = Shapes.join(
      Shapes.block(),
      Shapes.or(
          Block.box(0.0, 0.0, 4.0, 16.0, 3.0, 12.0),
          Block.box(4.0, 0.0, 0.0, 12.0, 3.0, 16.0),
          Block.box(2.0, 0.0, 2.0, 14.0, 3.0, 14.0),
          CrucibleBlock.CRUCIBLE_INSIDE
      ),
      BooleanOp.ONLY_FIRST
  );
  public static final BooleanProperty BOILING = T7BlockStateProperties.BOILING;

  public CrucibleBlock() {
    super(BlockBehaviour.Properties.ofFullCopy(Blocks.CAULDRON));
  }

  public static void meltItem(ServerLevel level, BlockPos pPos, ItemEntity itemEntity) {
    var thrownStack = itemEntity.getItem();

    // try to use as catalyst
    if (thrownStack.is(T7Tags.INSTANCE.getCATALYST())) {
      AspectMap crucibleAspects = AspectContainer
          .at(level, pPos).map(IAspectContainer::getAspects).orElseThrow();

      RecipeManager recipes = level.getRecipeManager();
      var input = new CrucibleRecipeInput(crucibleAspects, thrownStack);

      boolean success = recipes.getRecipeFor(
          T7RecipeTypes.INSTANCE.getCRUCIBLE().get(),
          input,
          level
      ).map(recipe -> {
        if (!CrucibleBlock.tryLowerFillLevel(level, pPos)) return false;
        CrucibleBlock.waterSplash(level, pPos);

        AspectContainer
            .at(level, pPos)
            .ifPresent(container ->
                container.extract(recipe.value().getRequiredAspects())
            );

        if (itemEntity.getOwner() instanceof ServerPlayer player) {
          ItemEntity itementity = player.drop(recipe.value().assemble(input, level.registryAccess()), true, true);
          if (itementity != null) {
            itementity.setNoPickUpDelay();
            itementity.setTarget(player.getUUID());
          }
          CrucibleBlock.shrinkItemEntity(itemEntity);
        }

        return true;
      }).orElse(false);

      if (success) return;// if catalyst failed, try to melt item instead
    }

    if (!AspectHelperKt.hasAspects(thrownStack) || !CrucibleBlock.hasWater(level, pPos)) return;
    AspectMap itemAspects = AspectHelperKt.getAspects(itemEntity);
    AspectContainer
        .at(level, pPos)
        .ifPresent(c -> c.insert(itemAspects));
    CrucibleBlock.waterSplash(level, pPos);
    itemEntity.discard();
  }

  private static void waterSplash(ServerLevel level, BlockPos pPos) {
    level.playSound(null, pPos, SoundEvents.GENERIC_SPLASH, SoundSource.BLOCKS, 1F, 1.0F);
  }

  // returns true if any water was drained
  public static boolean tryLowerFillLevel(Level pLevel, BlockPos pPos) {
    return pLevel.getBlockEntity(pPos, T7BlockEntities.INSTANCE.getCRUCIBLE().get()).map(be -> be.getFluidHandler().catalystDrain()).orElse(false);
  }

  public static boolean hasWater(Level pLevel, BlockPos pPos) {
    return pLevel.getBlockEntity(pPos, T7BlockEntities.INSTANCE.getCRUCIBLE().get()).map(be -> be.getFluidHandler().getFluidAmount() > 0).orElse(false);
  }

  // returns true if any water was filled
  public static boolean fillUp(Level pLevel, BlockPos pPos) {
    var be = pLevel.getBlockEntity(pPos);
    if (be instanceof CrucibleBE crucibleBE) return crucibleBE.getFluidHandler().fillUp();
    return false;
  }

  public static void shrinkItemEntity(ItemEntity itemEntity) {
    var stack = itemEntity.getItem();
    stack.shrink(1);
    if (stack.isEmpty()) itemEntity.discard();
    else itemEntity.setItem(stack.copy());
  }

  public static boolean isHeatSource(LevelAccessor level, BlockPos pos) {
    var bs = level.getBlockState(pos);
    var bsHeat = bs.is(T7Tags.CrucibleHeatSourceTag.INSTANCE.getBLOCK());
    var fs = level.getFluidState(pos);
    var fsHeat = fs.is(T7Tags.CrucibleHeatSourceTag.INSTANCE.getFLUID());
    return bsHeat || fsHeat;
  }

  @Override
  public void stepOn(Level pLevel, BlockPos pPos, BlockState pState, Entity pEntity) {
    if (pState.getValue(CrucibleBlock.BOILING) && !pEntity.isSteppingCarefully() && pEntity instanceof LivingEntity)
      pEntity.hurt(pLevel.damageSources().hotFloor(), 1.0F);

    super.stepOn(pLevel, pPos, pState, pEntity);
  }

  @Override
  @Nullable
  public BlockState getStateForPlacement(BlockPlaceContext pContext) {
    var below = pContext.getClickedPos().below();

    return this.defaultBlockState()
        .setValue(CrucibleBlock.BOILING, CrucibleBlock.isHeatSource(pContext.getLevel(), below));
  }

  @Override
  protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> pBuilder) {
    pBuilder.add(CrucibleBlock.BOILING);
  }

  protected boolean isEntityInsideContent(BlockPos pPos, Entity pEntity) {
    return pEntity.getY() < (double) pPos.getY() + 15 / 16f
        && pEntity.getBoundingBox().maxY > (double) pPos.getY() + 0.25;
  }

  @Nullable
  @Override
  public BlockEntity newBlockEntity(BlockPos pPos, BlockState pState) {
    return new CrucibleBE(pPos, pState);
  }

  @Nullable
  @Override
  public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level level, BlockState state, BlockEntityType<T> type) {
    return type == T7BlockEntities.INSTANCE.getCRUCIBLE().get() ? CrucibleBE::tick : null;
  }

  @Override
  protected boolean isPathfindable(BlockState pState, PathComputationType pPathComputationType) {
    return false;
  }

  @Override
  protected BlockState updateShape(BlockState pState, Direction pDirection, BlockState pNeighborState, LevelAccessor pLevel, BlockPos pPos, BlockPos pNeighborPos) {
    return pState.setValue(CrucibleBlock.BOILING, CrucibleBlock.isHeatSource(pLevel, pPos.below()));
  }

  @Override
  protected void onPlace(BlockState pState, Level pLevel, BlockPos pPos, BlockState pOldState, boolean pIsMoving) {
    super.onPlace(pState, pLevel, pPos, pOldState, pIsMoving);
    if (!pOldState.is(this)) pLevel.invalidateCapabilities(pPos);
  }

  @Override
  protected void onRemove(BlockState pState, Level pLevel, BlockPos pPos, BlockState pOldState, boolean pIsMoving) {
    super.onRemove(pState, pLevel, pPos, pOldState, pIsMoving);
    if (!pState.is(pOldState.getBlock())) pLevel.invalidateCapabilities(pPos);
  }

  @Override
  protected ItemInteractionResult useItemOn(ItemStack pStack, BlockState pState, Level pLevel, BlockPos pPos, Player pPlayer, InteractionHand pHand, BlockHitResult pHitResult) {
    // water buckets should be usable to top off, even if 1000 mB is too much
    if (pPlayer.getItemInHand(pHand).is(Items.WATER_BUCKET) && CrucibleBlock.fillUp(pLevel, pPos)) {
      if (!pPlayer.isCreative()) pPlayer.setItemInHand(pHand, new ItemStack(Items.BUCKET));
      pLevel.playSound(null, pPos, SoundEvents.BUCKET_EMPTY, SoundSource.BLOCKS, 1.0F, 1.0F);

      return ItemInteractionResult.SUCCESS;
    }

    // generic fluid interaction
    if (FluidUtil.interactWithFluidHandler(pPlayer, pHand, pLevel, pPos, pHitResult.getDirection()))
      return ItemInteractionResult.SUCCESS;

    return super.useItemOn(pStack, pState, pLevel, pPos, pPlayer, pHand, pHitResult);
  }

  @Override
  protected VoxelShape getInteractionShape(BlockState pState, BlockGetter pLevel, BlockPos pPos) {
    return CrucibleBlock.CRUCIBLE_INSIDE;
  }

  @Override
  protected VoxelShape getShape(BlockState pState, BlockGetter pLevel, BlockPos pPos, CollisionContext pContext) {
    return CrucibleBlock.CRUCIBLE_SHAPE;
  }

  @Override
  protected void entityInside(BlockState pState, Level level, BlockPos pPos, Entity pEntity) {
    if (!level.isClientSide
        && level instanceof ServerLevel serverLevel
        && pState.getValue(CrucibleBlock.BOILING)
        && pEntity instanceof ItemEntity itemEntity
        && this.isEntityInsideContent(pPos, pEntity)
    ) if (pEntity.mayInteract(level, pPos)) {
      CrucibleBlock.meltItem(serverLevel, pPos, itemEntity);
      level.sendBlockUpdated(
          pPos,
          pState,
          pState,
          Block.UPDATE_CLIENTS
      );
    }
  }
}

File: src/main/java/me/alegian/thavma/impl/common/block/CrucibleBlock.kt
package me.alegian.thavma.impl.common.block

import me.alegian.thavma.impl.common.aspect.getAspects
import me.alegian.thavma.impl.common.aspect.hasAspects
import me.alegian.thavma.impl.common.block.entity.CrucibleBE
import me.alegian.thavma.impl.common.data.capability.AspectContainer
import me.alegian.thavma.impl.common.data.capability.IAspectContainer
import me.alegian.thavma.impl.common.recipe.CrucibleRecipeInput
import me.alegian.thavma.impl.init.registries.T7Tags.CATALYST
import me.alegian.thavma.impl.init.registries.T7Tags.CrucibleHeatSourceTag.BLOCK
import me.alegian.thavma.impl.init.registries.T7Tags.CrucibleHeatSourceTag.FLUID
import me.alegian.thavma.impl.init.registries.deferred.T7BlockEntities
import me.alegian.thavma.impl.init.registries.deferred.T7RecipeTypes
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.sounds.SoundEvents
import net.minecraft.sounds.SoundSource
import net.minecraft.world.InteractionHand
import net.minecraft.world.ItemInteractionResult
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.item.ItemEntity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.item.context.BlockPlaceContext
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.Level
import net.minecraft.world.level.LevelAccessor
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.EntityBlock
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityTicker
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.StateDefinition
import net.minecraft.world.level.pathfinder.PathComputationType
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.shapes.BooleanOp
import net.minecraft.world.phys.shapes.CollisionContext
import net.minecraft.world.phys.shapes.Shapes
import net.minecraft.world.phys.shapes.VoxelShape
import net.neoforged.neoforge.fluids.FluidUtil

open class CrucibleBlock : Block(Properties.ofFullCopy(Blocks.CAULDRON)), EntityBlock {
    init {
        this.registerDefaultState(stateDefinition.any().setValue(BOILING, false))
    }

    override fun stepOn(pLevel: Level, pPos: BlockPos, pState: BlockState, pEntity: Entity) {
        if (pState.getValue(BOILING) && !pEntity.isSteppingCarefully && pEntity is LivingEntity)
            pEntity.hurt(pLevel.damageSources().hotFloor(), 1.0f)

        super.stepOn(pLevel, pPos, pState, pEntity)
    }

    override fun getStateForPlacement(pContext: BlockPlaceContext): BlockState? {
        return stateWithBoiling(pContext.level, pContext.clickedPos)
    }

    /**
     * Used by Wands to set BOILING state when converting cauldrons
     */
    open fun stateWithBoiling(level: LevelAccessor, blockPos: BlockPos): BlockState{
        val below = blockPos.below()

        return defaultBlockState().setValue(BOILING, isHeatSource(level, below))
    }

    override fun createBlockStateDefinition(pBuilder: StateDefinition.Builder<Block, BlockState>) {
        pBuilder.add(BOILING)
    }

    protected open fun isEntityInsideContent(pPos: BlockPos, pEntity: Entity): Boolean {
        return pEntity.y < pPos.y.toDouble() + 15 / 16f
                && pEntity.boundingBox.maxY > pPos.y.toDouble() + 0.25
    }

    override fun newBlockEntity(pPos: BlockPos, pState: BlockState): BlockEntity? {
        return CrucibleBE(pPos, pState)
    }

    override fun <T : BlockEntity?> getTicker(level: Level, state: BlockState, type: BlockEntityType<T>): BlockEntityTicker<T>? {
        return if (type === T7BlockEntities.CRUCIBLE.get()) BlockEntityTicker(CrucibleBE::tick) else null
    }

    override fun isPathfindable(pState: BlockState, pPathComputationType: PathComputationType): Boolean {
        return false
    }

    override fun updateShape(pState: BlockState, pDirection: Direction, pNeighborState: BlockState, pLevel: LevelAccessor, pPos: BlockPos, pNeighborPos: BlockPos): BlockState {
        return stateWithBoiling(pLevel, pPos)
    }

    override fun onPlace(pState: BlockState, pLevel: Level, pPos: BlockPos, pOldState: BlockState, pIsMoving: Boolean) {
        super.onPlace(pState, pLevel, pPos, pOldState, pIsMoving)
        if (!pOldState.`is`(this)) pLevel.invalidateCapabilities(pPos)
    }

    override fun onRemove(pState: BlockState, pLevel: Level, pPos: BlockPos, pOldState: BlockState, pIsMoving: Boolean) {
        super.onRemove(pState, pLevel, pPos, pOldState, pIsMoving)
        if (!pState.`is`(pOldState.block)) pLevel.invalidateCapabilities(pPos)
    }

    override fun useItemOn(pStack: ItemStack, pState: BlockState, pLevel: Level, pPos: BlockPos, pPlayer: Player, pHand: InteractionHand, pHitResult: BlockHitResult): ItemInteractionResult {
        // water buckets should be usable to top off, even if 1000 mB is too much
        if (pPlayer.getItemInHand(pHand).`is`(Items.WATER_BUCKET) && fillUp(pLevel, pPos)) {
            if (!pPlayer.isCreative) pPlayer.setItemInHand(pHand, ItemStack(Items.BUCKET))
            pLevel.playSound(null, pPos, SoundEvents.BUCKET_EMPTY, SoundSource.BLOCKS, 1.0f, 1.0f)

            return ItemInteractionResult.SUCCESS
        }

        // generic fluid interaction
        if (FluidUtil.interactWithFluidHandler(pPlayer, pHand, pLevel, pPos, pHitResult.direction))
            return ItemInteractionResult.SUCCESS

        return super.useItemOn(pStack, pState, pLevel, pPos, pPlayer, pHand, pHitResult)
    }

    override fun getInteractionShape(pState: BlockState, pLevel: BlockGetter, pPos: BlockPos): VoxelShape {
        return CRUCIBLE_INSIDE
    }

    override fun getShape(pState: BlockState, pLevel: BlockGetter, pPos: BlockPos, pContext: CollisionContext): VoxelShape {
        return CRUCIBLE_SHAPE
    }

    override fun entityInside(pState: BlockState, level: Level, pPos: BlockPos, pEntity: Entity) {
        if (!level.isClientSide && level is ServerLevel
            && pState.getValue(BOILING)
            && pEntity is ItemEntity
            && this.isEntityInsideContent(pPos, pEntity)
            && pEntity.mayInteract(level, pPos)
        ) {
            meltItem(level, pPos, pEntity)
            level.sendBlockUpdated(
                pPos,
                pState,
                pState,
                UPDATE_CLIENTS
            )
        }
    }

    companion object {
        val CRUCIBLE_INSIDE = box(2.0, 4.0, 2.0, 14.0, 16.0, 14.0)
        val CRUCIBLE_SHAPE = Shapes.join(
            Shapes.block(),
            Shapes.or(
                box(0.0, 0.0, 4.0, 16.0, 3.0, 12.0),
                box(4.0, 0.0, 0.0, 12.0, 3.0, 16.0),
                box(2.0, 0.0, 2.0, 14.0, 3.0, 14.0),
                CRUCIBLE_INSIDE
            ),
            BooleanOp.ONLY_FIRST
        )
        val BOILING = T7BlockStateProperties.BOILING

        protected fun meltItem(level: ServerLevel, pPos: BlockPos, itemEntity: ItemEntity) {
            val thrownStack = itemEntity.item

            // try to use as catalyst
            if (thrownStack.`is`(CATALYST)) {
                val crucibleAspects = AspectContainer.at(level, pPos).map(IAspectContainer::getAspects).orElseThrow()

                val recipes = level.recipeManager
                val input = CrucibleRecipeInput(crucibleAspects, thrownStack)

                val success = recipes.getRecipeFor(T7RecipeTypes.CRUCIBLE.get(), input, level).map { recipe ->
                    if (!tryLowerFillLevel(level, pPos)) return@map false
                    waterSplash(level, pPos)

                    AspectContainer.at(level, pPos).ifPresent { container ->
                        container.extract(recipe.value().requiredAspects)
                    }

                    val player = itemEntity.owner
                    if (player is ServerPlayer) {
                        player.drop(recipe.value().assemble(input, level.registryAccess()), true, true)?.run {
                            setNoPickUpDelay()
                            target = player.getUUID()
                            shrink()
                        }
                    }
                    true
                }.orElse(false)

                if (success) return  // if catalyst failed, try to melt item instead
            }

            if (!hasAspects(thrownStack) || !hasWater(level, pPos)) return
            val itemAspects = getAspects(itemEntity)
            AspectContainer.at(level, pPos).ifPresent { c -> c.insert(itemAspects) }
            waterSplash(level, pPos)
            itemEntity.discard()
        }

        protected fun waterSplash(level: ServerLevel, pPos: BlockPos) {
            level.playSound(null, pPos, SoundEvents.GENERIC_SPLASH, SoundSource.BLOCKS, 1f, 1.0f)
        }

        // returns true if any water was drained
        protected fun tryLowerFillLevel(pLevel: Level, pPos: BlockPos): Boolean {
            return pLevel.getBlockEntity(pPos, T7BlockEntities.CRUCIBLE.get()).map { be -> be.fluidHandler.catalystDrain() }.orElse(false)
        }

        protected fun hasWater(pLevel: Level, pPos: BlockPos): Boolean {
            return pLevel.getBlockEntity(pPos, T7BlockEntities.CRUCIBLE.get()).map { be -> be.fluidHandler.fluidAmount > 0 }.orElse(false)
        }

        // returns true if any water was filled
        protected fun fillUp(pLevel: Level, pPos: BlockPos): Boolean {
            val be = pLevel.getBlockEntity(pPos)
            if (be is CrucibleBE) return be.fluidHandler.fillUp()
            return false
        }

        protected fun ItemEntity.shrink() {
            item.shrink(1)
            if (item.isEmpty) discard()
            else item = item.copy()
        }

        protected fun isHeatSource(level: LevelAccessor, pos: BlockPos): Boolean {
            val bs = level.getBlockState(pos)
            val bsHeat = bs.`is`(BLOCK)
            val fs = level.getFluidState(pos)
            val fsHeat = fs.`is`(FLUID)
            return bsHeat || fsHeat
        }
    }
}
